File: test/priority_arbitrator.cpp
#include <memory>
#include <string>

#include <gtest/gtest.h>

#include "behavior.hpp"
#include "priority_arbitrator.hpp"

#include "dummy_types.hpp"


using namespace arbitration_graphs;
using namespace arbitration_graphs_tests;


class PriorityArbitratorTest : public ::testing::Test {
protected:
    using OptionFlags = PriorityArbitrator<DummyEnvironmentModel, DummyCommand>::Option::Flags;

    DummyBehavior::Ptr testBehaviorHighPriority = std::make_shared<DummyBehavior>(false, false, "HighPriority");
    DummyBehavior::Ptr testBehaviorMidPriority = std::make_shared<DummyBehavior>(true, false, "MidPriority");
    DummyBehavior::Ptr testBehaviorLowPriority = std::make_shared<DummyBehavior>(true, true, "LowPriority");


    PriorityArbitrator<DummyEnvironmentModel, DummyCommand> testPriorityArbitrator;

    DummyEnvironmentModel environmentModel;

    Time time{Clock::now()};
};


TEST_F(PriorityArbitratorTest, BasicFunctionality) {
    DummyEnvironmentModel environmentModel;

    // if there are no options yet -> the invocationCondition should be false
    EXPECT_FALSE(testPriorityArbitrator.checkInvocationCondition(time, environmentModel));
    EXPECT_FALSE(testPriorityArbitrator.checkCommitmentCondition(time, environmentModel));

    // otherwise the invocationCondition is true if any of the option has true invocationCondition
    testPriorityArbitrator.addOption(testBehaviorHighPriority, OptionFlags::NoFlags);
    testPriorityArbitrator.addOption(testBehaviorHighPriority, OptionFlags::NoFlags);
    EXPECT_FALSE(testPriorityArbitrator.checkInvocationCondition(time, environmentModel));
    EXPECT_FALSE(testPriorityArbitrator.checkCommitmentCondition(time, environmentModel));

    testPriorityArbitrator.addOption(testBehaviorMidPriority, OptionFlags::NoFlags);
    testPriorityArbitrator.addOption(testBehaviorLowPriority, OptionFlags::NoFlags);

    EXPECT_TRUE(testPriorityArbitrator.checkInvocationCondition(time, environmentModel));
    EXPECT_FALSE(testPriorityArbitrator.checkCommitmentCondition(time, environmentModel));

    testPriorityArbitrator.gainControl(time, environmentModel);

    EXPECT_EQ("MidPriority", testPriorityArbitrator.getCommand(time, environmentModel));
    EXPECT_EQ(0, testBehaviorMidPriority->loseControlCounter_);

    // testBehaviorMidPriority.loseControl(time) should be called within getCommand since commitment condition is false
    EXPECT_EQ("MidPriority", testPriorityArbitrator.getCommand(time, environmentModel));
    EXPECT_EQ(1, testBehaviorMidPriority->loseControlCounter_);

    testBehaviorMidPriority->invocationCondition_ = false;
    EXPECT_TRUE(testPriorityArbitrator.checkInvocationCondition(time, environmentModel));
    EXPECT_TRUE(testPriorityArbitrator.checkCommitmentCondition(time, environmentModel));

    EXPECT_EQ("LowPriority", testPriorityArbitrator.getCommand(time, environmentModel));
    EXPECT_EQ(0, testBehaviorLowPriority->loseControlCounter_);

    // testBehaviorLowPriority.loseControl(time) should NOT be called within getCommand since commitment condition is
    // true
    EXPECT_EQ("LowPriority", testPriorityArbitrator.getCommand(time, environmentModel));
    EXPECT_EQ(0, testBehaviorLowPriority->loseControlCounter_);

    testBehaviorMidPriority->invocationCondition_ = true;
    EXPECT_TRUE(testPriorityArbitrator.checkInvocationCondition(time, environmentModel));
    EXPECT_TRUE(testPriorityArbitrator.checkCommitmentCondition(time, environmentModel));
    EXPECT_EQ("LowPriority", testPriorityArbitrator.getCommand(time, environmentModel));
    EXPECT_EQ("LowPriority", testPriorityArbitrator.getCommand(time, environmentModel));
}

TEST_F(PriorityArbitratorTest, Printout) {
    DummyEnvironmentModel environmentModel;

    testPriorityArbitrator.addOption(testBehaviorHighPriority, OptionFlags::NoFlags);
    testPriorityArbitrator.addOption(testBehaviorHighPriority, OptionFlags::NoFlags);
    testPriorityArbitrator.addOption(testBehaviorMidPriority, OptionFlags::NoFlags);
    testPriorityArbitrator.addOption(testBehaviorLowPriority, OptionFlags::NoFlags);

    // clang-format off
    std::string expected_printout = invocationTrueString + commitmentFalseString + "PriorityArbitrator\n"
                                    "    1. " + invocationFalseString + commitmentFalseString + "HighPriority\n"
                                    "    2. " + invocationFalseString + commitmentFalseString + "HighPriority\n"
                                    "    3. " + invocationTrueString + commitmentFalseString + "MidPriority\n"
                                    "    4. " + invocationTrueString + commitmentTrueString + "LowPriority";
    // clang-format on
    std::string actual_printout = testPriorityArbitrator.to_str(time, environmentModel);
    std::cout << actual_printout << std::endl;

    EXPECT_EQ(expected_printout, actual_printout);


    testPriorityArbitrator.gainControl(time, environmentModel);
    EXPECT_EQ("MidPriority", testPriorityArbitrator.getCommand(time, environmentModel));

    // clang-format off
    expected_printout = invocationTrueString + commitmentTrueString + "PriorityArbitrator\n"
                        "    1. " + invocationFalseString + commitmentFalseString + "HighPriority\n"
                        "    2. " + invocationFalseString + commitmentFalseString + "HighPriority\n"
                        " -> 3. " + invocationTrueString + commitmentFalseString + "MidPriority\n"
                        "    4. " + invocationTrueString + commitmentTrueString + "LowPriority";
    // clang-format on
    actual_printout = testPriorityArbitrator.to_str(time, environmentModel);
    std::cout << actual_printout << std::endl;

    EXPECT_EQ(expected_printout, actual_printout);
}

TEST_F(PriorityArbitratorTest, ToYaml) {
    DummyEnvironmentModel environmentModel;

    testPriorityArbitrator.addOption(testBehaviorHighPriority, OptionFlags::NoFlags);
    testPriorityArbitrator.addOption(testBehaviorHighPriority, OptionFlags::NoFlags);
    testPriorityArbitrator.addOption(testBehaviorMidPriority, OptionFlags::Interruptable);
    testPriorityArbitrator.addOption(testBehaviorLowPriority, OptionFlags::NoFlags);

    YAML::Node yaml = testPriorityArbitrator.toYaml(time, environmentModel);

    //    std::cout << yaml << std::endl << std::endl;

    EXPECT_EQ("PriorityArbitrator", yaml["type"].as<std::string>());
    EXPECT_EQ("PriorityArbitrator", yaml["name"].as<std::string>());
    EXPECT_EQ(true, yaml["invocationCondition"].as<bool>());
    EXPECT_EQ(false, yaml["commitmentCondition"].as<bool>());

    ASSERT_EQ(4, yaml["options"].size());
    EXPECT_EQ("Option", yaml["options"][0]["type"].as<std::string>());
    EXPECT_EQ("Option", yaml["options"][1]["type"].as<std::string>());
    EXPECT_EQ("Option", yaml["options"][2]["type"].as<std::string>());
    EXPECT_EQ("Option", yaml["options"][3]["type"].as<std::string>());
    EXPECT_EQ("HighPriority", yaml["options"][0]["behavior"]["name"].as<std::string>());
    EXPECT_EQ("HighPriority", yaml["options"][1]["behavior"]["name"].as<std::string>());
    EXPECT_EQ("MidPriority", yaml["options"][2]["behavior"]["name"].as<std::string>());
    EXPECT_EQ("LowPriority", yaml["options"][3]["behavior"]["name"].as<std::string>());
    EXPECT_EQ(false, yaml["options"][0]["flags"].IsDefined());
    EXPECT_EQ(false, yaml["options"][1]["flags"].IsDefined());
    ASSERT_EQ(true, yaml["options"][2]["flags"].IsDefined());
    EXPECT_EQ(false, yaml["options"][3]["flags"].IsDefined());

    EXPECT_EQ(false, yaml["activeBehavior"].IsDefined());

    testPriorityArbitrator.gainControl(time, environmentModel);
    testPriorityArbitrator.getCommand(time, environmentModel);

    yaml = testPriorityArbitrator.toYaml(time, environmentModel);

    EXPECT_EQ(true, yaml["invocationCondition"].as<bool>());
    EXPECT_EQ(true, yaml["commitmentCondition"].as<bool>());

    ASSERT_EQ(true, yaml["activeBehavior"].IsDefined());
    EXPECT_EQ(2, yaml["activeBehavior"].as<int>());
}


TEST_F(PriorityArbitratorTest, BasicFunctionalityWithInterruptableOptions) {
    // if there are no options yet -> the invocationCondition should be false
    EXPECT_FALSE(testPriorityArbitrator.checkInvocationCondition(time, environmentModel));
    EXPECT_FALSE(testPriorityArbitrator.checkCommitmentCondition(time, environmentModel));

    // otherwise the invocationCondition is true if any of the option has true invocationCondition
    testPriorityArbitrator.addOption(testBehaviorHighPriority, OptionFlags::Interruptable);
    testPriorityArbitrator.addOption(testBehaviorHighPriority, OptionFlags::Interruptable);
    EXPECT_FALSE(testPriorityArbitrator.checkInvocationCondition(time, environmentModel));
    EXPECT_FALSE(testPriorityArbitrator.checkCommitmentCondition(time, environmentModel));

    testPriorityArbitrator.addOption(testBehaviorMidPriority, OptionFlags::Interruptable);
    testPriorityArbitrator.addOption(testBehaviorLowPriority, OptionFlags::Interruptable);

    EXPECT_TRUE(testPriorityArbitrator.checkInvocationCondition(time, environmentModel));
    EXPECT_FALSE(testPriorityArbitrator.checkCommitmentCondition(time, environmentModel));

    testPriorityArbitrator.gainControl(time, environmentModel);
    EXPECT_EQ("MidPriority", testPriorityArbitrator.getCommand(time, environmentModel));
    EXPECT_EQ("MidPriority", testPriorityArbitrator.getCommand(time, environmentModel));

    testBehaviorMidPriority->invocationCondition_ = false;
    EXPECT_TRUE(testPriorityArbitrator.checkInvocationCondition(time, environmentModel));
    EXPECT_TRUE(testPriorityArbitrator.checkCommitmentCondition(time, environmentModel));
    EXPECT_EQ("LowPriority", testPriorityArbitrator.getCommand(time, environmentModel));
    EXPECT_EQ("LowPriority", testPriorityArbitrator.getCommand(time, environmentModel));

    testBehaviorMidPriority->invocationCondition_ = true;
    EXPECT_TRUE(testPriorityArbitrator.checkInvocationCondition(time, environmentModel));
    EXPECT_TRUE(testPriorityArbitrator.checkCommitmentCondition(time, environmentModel));
    EXPECT_EQ("MidPriority", testPriorityArbitrator.getCommand(time, environmentModel));
    EXPECT_EQ("MidPriority", testPriorityArbitrator.getCommand(time, environmentModel));
}

TEST(PriorityArbitrator, SubCommandTypeDiffersFromCommandType) {
    Time time{Clock::now()};
    DummyEnvironmentModel environmentModel;

    using OptionFlags = PriorityArbitrator<DummyEnvironmentModel, DummyCommandInt, DummyCommand>::Option::Flags;

    DummyBehavior::Ptr testBehaviorHighPriority = std::make_shared<DummyBehavior>(false, false, "___HighPriority___");
    DummyBehavior::Ptr testBehaviorMidPriority = std::make_shared<DummyBehavior>(true, false, "__MidPriority__");
    DummyBehavior::Ptr testBehaviorLowPriority = std::make_shared<DummyBehavior>(true, true, "_LowPriority_");

    PriorityArbitrator<DummyEnvironmentModel, DummyCommandInt, DummyCommand> testPriorityArbitrator;

    testPriorityArbitrator.addOption(testBehaviorHighPriority, OptionFlags::NoFlags);
    testPriorityArbitrator.addOption(testBehaviorHighPriority, OptionFlags::NoFlags);
    testPriorityArbitrator.addOption(testBehaviorMidPriority, OptionFlags::NoFlags);
    testPriorityArbitrator.addOption(testBehaviorLowPriority, OptionFlags::NoFlags);

    testPriorityArbitrator.gainControl(time, environmentModel);

    std::string expected = "__MidPriority__";
    EXPECT_EQ(expected.length(), testPriorityArbitrator.getCommand(time, environmentModel));
}

File: test/priority_arbitrator.py
import sys
import os

build_directory = "/home/blinky/arbitration_graphs/build/"
sys.path.append(build_directory)

import time
import unittest

import yaml

import arbitration_graphs_py as ag
from dummy_types import DummyBehaviorPy, DummyCommandPy


class TestPriorityArbitrator(unittest.TestCase):
    def setUp(self):
        self.test_behavior_high_priority = DummyBehaviorPy(False, False, "HighPriority")
        self.test_behavior_mid_priority = DummyBehaviorPy(True, False, "MidPriority")
        self.test_behavior_low_priority = DummyBehaviorPy(True, True, "LowPriority")

        self.test_priority_arbitrator = ag.PriorityArbitrator("PriorityArbitrator")

        self.time = time.time()

    def test_basic_functionality(self):
        # Test if there are no options -> invocationCondition should be False
        self.assertFalse(
            self.test_priority_arbitrator.check_invocation_condition(self.time)
        )
        self.assertFalse(
            self.test_priority_arbitrator.check_commitment_condition(self.time)
        )

        # Adding options but no invocation
        self.test_priority_arbitrator.add_option(
            self.test_behavior_high_priority,
            ag.PriorityArbitrator.Option.Flags.NO_FLAGS,
        )
        self.test_priority_arbitrator.add_option(
            self.test_behavior_high_priority,
            ag.PriorityArbitrator.Option.Flags.NO_FLAGS,
        )
        self.assertFalse(
            self.test_priority_arbitrator.check_invocation_condition(self.time)
        )
        self.assertFalse(
            self.test_priority_arbitrator.check_commitment_condition(self.time)
        )

        # Adding mid and low priority options
        self.test_priority_arbitrator.add_option(
            self.test_behavior_mid_priority, ag.PriorityArbitrator.Option.Flags.NO_FLAGS
        )
        self.test_priority_arbitrator.add_option(
            self.test_behavior_low_priority, ag.PriorityArbitrator.Option.Flags.NO_FLAGS
        )
        # Invocation should be true but not commitment
        self.assertTrue(
            self.test_priority_arbitrator.check_invocation_condition(self.time)
        )
        self.assertFalse(
            self.test_priority_arbitrator.check_commitment_condition(self.time)
        )

        # Gain control and check command
        self.test_priority_arbitrator.gain_control(self.time)
        self.assertEqual(
            "MidPriority", self.test_priority_arbitrator.get_command(self.time)
        )
        self.assertEqual(0, self.test_behavior_mid_priority.lose_control_counter)

        # Mid priority behavior should lose control after the next get_command call
        self.assertEqual(
            "MidPriority", self.test_priority_arbitrator.get_command(self.time)
        )
        self.assertEqual(1, self.test_behavior_mid_priority.lose_control_counter)

        # Now set mid-priority invocation to false and check again
        self.test_behavior_mid_priority.invocation_condition = False
        self.assertTrue(
            self.test_priority_arbitrator.check_invocation_condition(self.time)
        )
        self.assertTrue(
            self.test_priority_arbitrator.check_commitment_condition(self.time)
        )
        self.assertEqual(
            "LowPriority", self.test_priority_arbitrator.get_command(self.time)
        )
        self.assertEqual(0, self.test_behavior_low_priority.lose_control_counter)

        # Low priority behavior should not lose control if commitment is true
        self.assertEqual(
            "LowPriority", self.test_priority_arbitrator.get_command(self.time)
        )
        self.assertEqual(0, self.test_behavior_low_priority.lose_control_counter)

        # Test with mid-priority invocation set to true again
        self.test_behavior_mid_priority.invocation_condition = True
        self.assertTrue(
            self.test_priority_arbitrator.check_invocation_condition(self.time)
        )
        self.assertTrue(
            self.test_priority_arbitrator.check_commitment_condition(self.time)
        )
        self.assertEqual(
            "LowPriority", self.test_priority_arbitrator.get_command(self.time)
        )
        self.assertEqual(
            "LowPriority", self.test_priority_arbitrator.get_command(self.time)
        )

    def test_printout(self):
        # Add options to the arbitrator
        self.test_priority_arbitrator.add_option(
            self.test_behavior_high_priority,
            ag.PriorityArbitrator.Option.Flags.NO_FLAGS,
        )
        self.test_priority_arbitrator.add_option(
            self.test_behavior_high_priority,
            ag.PriorityArbitrator.Option.Flags.NO_FLAGS,
        )
        self.test_priority_arbitrator.add_option(
            self.test_behavior_mid_priority, ag.PriorityArbitrator.Option.Flags.NO_FLAGS
        )
        self.test_priority_arbitrator.add_option(
            self.test_behavior_low_priority, ag.PriorityArbitrator.Option.Flags.NO_FLAGS
        )

        # fmt: off
        expected_printout = (
            ag.testing_types.invocation_true_string + ag.testing_types.commitment_false_string + "PriorityArbitrator\n"
            "    1. " + ag.testing_types.invocation_false_string + ag.testing_types.commitment_false_string + "HighPriority\n"
            "    2. " + ag.testing_types.invocation_false_string + ag.testing_types.commitment_false_string + "HighPriority\n"
            "    3. " + ag.testing_types.invocation_true_string + ag.testing_types.commitment_false_string + "MidPriority\n"
            "    4. " + ag.testing_types.invocation_true_string + ag.testing_types.commitment_true_string + "LowPriority"
        )
        # fmt: on
        actual_printout = self.test_priority_arbitrator.to_str(self.time)
        print(actual_printout)

        self.assertEqual(expected_printout, actual_printout)

        self.test_priority_arbitrator.gain_control(self.time)
        self.assertEqual(
            "MidPriority", self.test_priority_arbitrator.get_command(self.time)
        )

        # fmt: off
        expected_printout = (
            ag.testing_types.invocation_true_string + ag.testing_types.commitment_true_string + "PriorityArbitrator\n"
            "    1. " + ag.testing_types.invocation_false_string + ag.testing_types.commitment_false_string + "HighPriority\n"
            "    2. " + ag.testing_types.invocation_false_string + ag.testing_types.commitment_false_string + "HighPriority\n"
            " -> 3. " + ag.testing_types.invocation_true_string + ag.testing_types.commitment_false_string + "MidPriority\n"
            "    4. " + ag.testing_types.invocation_true_string + ag.testing_types.commitment_true_string + "LowPriority"
        )
        # fmt: on
        actual_printout = self.test_priority_arbitrator.to_str(self.time)
        print(actual_printout)

        self.assertEqual(expected_printout, actual_printout)

    def test_to_yaml(self):
        # Add options to the arbitrator
        self.test_priority_arbitrator.add_option(
            self.test_behavior_high_priority,
            ag.PriorityArbitrator.Option.Flags.NO_FLAGS,
        )
        self.test_priority_arbitrator.add_option(
            self.test_behavior_high_priority,
            ag.PriorityArbitrator.Option.Flags.NO_FLAGS,
        )
        self.test_priority_arbitrator.add_option(
            self.test_behavior_mid_priority,
            ag.PriorityArbitrator.Option.Flags.INTERRUPTABLE,
        )
        self.test_priority_arbitrator.add_option(
            self.test_behavior_low_priority, ag.PriorityArbitrator.Option.Flags.NO_FLAGS
        )

        yaml_node = yaml.safe_load(
            self.test_priority_arbitrator.to_yaml_as_str(self.time)
        )

        self.assertEqual("PriorityArbitrator", yaml_node["type"])
        self.assertEqual("PriorityArbitrator", yaml_node["name"])
        self.assertTrue(yaml_node["invocationCondition"])
        self.assertFalse(yaml_node["commitmentCondition"])

        self.assertEqual(4, len(yaml_node["options"]))
        self.assertEqual("Option", yaml_node["options"][0]["type"])
        self.assertEqual("Option", yaml_node["options"][1]["type"])
        self.assertEqual("Option", yaml_node["options"][2]["type"])
        self.assertEqual("Option", yaml_node["options"][3]["type"])
        self.assertEqual("HighPriority", yaml_node["options"][0]["behavior"]["name"])
        self.assertEqual("HighPriority", yaml_node["options"][1]["behavior"]["name"])
        self.assertEqual("MidPriority", yaml_node["options"][2]["behavior"]["name"])
        self.assertEqual("LowPriority", yaml_node["options"][3]["behavior"]["name"])
        self.assertFalse("flags" in yaml_node["options"][0])
        self.assertFalse("flags" in yaml_node["options"][1])
        self.assertTrue("flags" in yaml_node["options"][2])
        self.assertFalse("flags" in yaml_node["options"][3])

        self.assertFalse("activeBehavior" in yaml_node)

        self.test_priority_arbitrator.gain_control(self.time)
        self.test_priority_arbitrator.get_command(self.time)

        yaml_node = yaml.safe_load(
            self.test_priority_arbitrator.to_yaml_as_str(self.time)
        )

        self.assertTrue(yaml_node["invocationCondition"])
        self.assertTrue(yaml_node["commitmentCondition"])

        self.assertTrue("activeBehavior" in yaml_node)
        self.assertEqual(2, yaml_node["activeBehavior"])

    def test_basic_functionality_with_interruptable_options(self):
        # if there are no options yet -> the invocationCondition should be false
        self.assertFalse(
            self.test_priority_arbitrator.check_invocation_condition(self.time)
        )
        self.assertFalse(
            self.test_priority_arbitrator.check_commitment_condition(self.time)
        )

        # otherwise the invocationCondition is true if any of the option has true invocationCondition
        self.test_priority_arbitrator.add_option(
            self.test_behavior_high_priority,
            ag.PriorityArbitrator.Option.Flags.INTERRUPTABLE,
        )
        self.test_priority_arbitrator.add_option(
            self.test_behavior_high_priority,
            ag.PriorityArbitrator.Option.Flags.INTERRUPTABLE,
        )
        self.assertFalse(
            self.test_priority_arbitrator.check_invocation_condition(self.time)
        )
        self.assertFalse(
            self.test_priority_arbitrator.check_commitment_condition(self.time)
        )

        self.test_priority_arbitrator.add_option(
            self.test_behavior_mid_priority,
            ag.PriorityArbitrator.Option.Flags.INTERRUPTABLE,
        )
        self.test_priority_arbitrator.add_option(
            self.test_behavior_low_priority,
            ag.PriorityArbitrator.Option.Flags.INTERRUPTABLE,
        )

        self.assertTrue(
            self.test_priority_arbitrator.check_invocation_condition(self.time)
        )
        self.assertFalse(
            self.test_priority_arbitrator.check_commitment_condition(self.time)
        )

        self.test_priority_arbitrator.gain_control(self.time)
        self.assertEqual(
            "MidPriority", self.test_priority_arbitrator.get_command(self.time)
        )
        self.assertEqual(
            "MidPriority", self.test_priority_arbitrator.get_command(self.time)
        )

        self.test_behavior_mid_priority.invocation_condition = False
        self.assertTrue(
            self.test_priority_arbitrator.check_invocation_condition(self.time)
        )
        self.assertTrue(
            self.test_priority_arbitrator.check_commitment_condition(self.time)
        )
        self.assertEqual(
            "LowPriority", self.test_priority_arbitrator.get_command(self.time)
        )
        self.assertEqual(
            "LowPriority", self.test_priority_arbitrator.get_command(self.time)
        )

        self.test_behavior_mid_priority.invocation_condition = True
        self.assertTrue(
            self.test_priority_arbitrator.check_invocation_condition(self.time)
        )
        self.assertTrue(
            self.test_priority_arbitrator.check_commitment_condition(self.time)
        )
        self.assertEqual(
            "MidPriority", self.test_priority_arbitrator.get_command(self.time)
        )
        self.assertEqual(
            "MidPriority", self.test_priority_arbitrator.get_command(self.time)
        )

    # def test_sub_command_type_differs_from_command_type(self):
    #     # Different sub-command type case
    #     test_priority_arbitrator = PriorityArbitrator("test_arbitrator_with_subcommand")
    #     test_behavior_high_priority = DummyBehavior(False, False, "___HighPriority___")
    #     test_behavior_mid_priority = DummyBehavior(True, False, "__MidPriority__")
    #     test_behavior_low_priority = DummyBehavior(True, True, "_LowPriority_")
    #
    #     test_priority_arbitrator.add_option(
    #         test_behavior_high_priority, ag.PriorityArbitrator.Option.Flags.NO_FLAGS
    #     )
    #     test_priority_arbitrator.add_option(
    #         test_behavior_high_priority, ag.PriorityArbitrator.Option.Flags.NO_FLAGS
    #     )
    #     test_priority_arbitrator.add_option(
    #         test_behavior_mid_priority, ag.PriorityArbitrator.Option.Flags.NO_FLAGS
    #     )
    #     test_priority_arbitrator.add_option(
    #         test_behavior_low_priority, ag.PriorityArbitrator.Option.Flags.NO_FLAGS
    #     )
    #
    #     test_priority_arbitrator.gain_control(self.time)
    #
    #     expected = "__MidPriority__"
    #     self.assertEqual(len(expected), test_priority_arbitrator.get_command(self.time))


if __name__ == "__main__":
    unittest.main()
